From b0431b04065b9cc66a139f1f42c456779a800c05 Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Fri, 9 Mar 2018 14:12:58 +0200
Subject: [PATCH 3474/5242] MLK-17648-2: drm: imx: dcss: Load the HDR10 from
 header file

commit  5cabd42b345565ff06fc888d641219e3b231720e from
https://source.codeaurora.org/external/imx/linux-imx.git

This commit allows one to select if a firmware file is used, for loading
the HDR10 tables, or a header. By default, this will be header file.
This is until a proper way of passing the file from bootloader is found.

Also, fix a minor bug which made parsing the tables over the actual data
limit.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dcss/dcss-hdr10.c |   17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/imx/dcss/dcss-hdr10.c b/drivers/gpu/imx/dcss/dcss-hdr10.c
index aa60cd7..18a6615 100644
--- a/drivers/gpu/imx/dcss/dcss-hdr10.c
+++ b/drivers/gpu/imx/dcss/dcss-hdr10.c
@@ -21,6 +21,12 @@
 #include <video/imx-dcss.h>
 #include "dcss-prv.h"
 
+#define USE_TBL_HEADER
+
+#ifdef USE_TBL_HEADER
+#include "dcss-hdr10-tables.h"
+#endif
+
 #define USE_CTXLD
 
 #define DCSS_HDR10_A0_LUT		0x0000
@@ -443,10 +449,11 @@ static void dcss_hdr10_parse_fw_data(struct dcss_hdr10_priv *hdr10)
 		dcss_hdr10_tbl_add(hdr10, tbl_desc, tbl_size, data);
 
 		data += tbl_size;
-		remaining -= tbl_size + 2;
+		remaining -= tbl_size + 3;
 	}
 }
 
+#ifndef USE_TBL_HEADER
 static void dcss_hdr10_fw_handler(const struct firmware *fw, void *context)
 {
 	struct dcss_hdr10_priv *hdr10 = context;
@@ -487,6 +494,7 @@ static void dcss_hdr10_fw_handler(const struct firmware *fw, void *context)
 
 	dev_info(hdr10->dcss->dev, "hdr10: DCSS FW loaded successfully\n");
 }
+#endif
 
 static int dcss_hdr10_tbls_init(struct dcss_hdr10_priv *hdr10)
 {
@@ -530,6 +538,7 @@ int dcss_hdr10_init(struct dcss_soc *dcss, unsigned long hdr10_base)
 		return ret;
 	}
 
+#ifndef USE_TBL_HEADER
 	ret = request_firmware_nowait(THIS_MODULE, FW_ACTION_HOTPLUG, "dcss.fw",
 				      dcss->dev, GFP_KERNEL, hdr10,
 				      dcss_hdr10_fw_handler);
@@ -537,6 +546,12 @@ int dcss_hdr10_init(struct dcss_soc *dcss, unsigned long hdr10_base)
 		dev_err(dcss->dev, "hdr10: Cannot async load DCSS FW.\n");
 		return ret;
 	}
+#else
+	hdr10->fw_data = (u8 *)dcss_hdr10_tables;
+	hdr10->fw_size = sizeof(dcss_hdr10_tables);
+
+	dcss_hdr10_parse_fw_data(hdr10);
+#endif
 
 	return dcss_hdr10_ch_init_all(dcss, hdr10_base);
 }
-- 
1.7.9.5

